### PR TITLE
gko-1984: REST model and repository layer for custom dashboards

### PR DESCRIPTION
**Pull request by a-cordier**
see https://gravitee.atlassian.net/browse/GKO-1984
specifically https://gravitee.atlassian.net/browse/GKO-2317

### CODE REVIEW

File: gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/resources/liquibase/changelogs/v4_11_0/06_create_custom_dashboards_table.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+            constraintName: pk_${gravitee_prefix}custom_dashboards
+            columnNames: id
+            tableName: ${gravitee_prefix}custom_dashboards
+        - createIndex:
+            tableName: ${gravitee_prefix}custom_dashboards
+            indexName: idx_${gravitee_prefix}custom_dashboards_organization_id
+            columns:
+              - column:
+                  name: organization_id
+                  type: nvarchar(64)
```
[the remark this answers is not in the document]
Comment (a-cordier) [reply gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/resources/liquibase/changelogs/v4_11_0/06_create_custom_dashboards_table.yml]: Done

File: gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/MongoCustomDashboardRepository.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        var res = mapper.map(createdDashboardMongo);
+        log.debug("Create custom dashboard [{}] - Done", customDashboard.getName());
+        return res;
+    }
+
+    @Override
+    public CustomDashboard update(CustomDashboard customDashboard) throws TechnicalException {
+        if (customDashboard == null || customDashboard.getName() == null) {
+            throw new IllegalStateException("Custom dashboard to update must have a name");
+        }
```
[the remark this answers is not in the document]
Comment (a-cordier) [reply gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/MongoCustomDashboardRepository.java:73]: Service layer is coming in another PR but 👍

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-analytics.yaml [hunk 3]
```diff
@@ -1230,6 +1377,269 @@ components:
...
+                    type: integer
+                    minimum: 0
+                    description: Vertical position (row index) in the grid
+                    example: 0
+            example:
+                cols: 1
+                rows: 1
+                x: 0
+                "y": 0
+
```
[the remark this answers is not in the document]
Comment (a-cordier) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-analytics.yaml:1422]: Done

File: gravitee-apim-repository/gravitee-apim-repository-api/src/main/java/io/gravitee/repository/management/model/CustomDashboardWidget.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    @Builder
+    @NoArgsConstructor
+    @AllArgsConstructor
+    @Data
+    public static class Request {
+
+        private String type;
+        private String timeRange;
```
Comment (carlos-andres-osorio) [inline gravitee-apim-repository/gravitee-apim-repository-api/src/main/java/io/gravitee/repository/management/model/CustomDashboardWidget.java]: just out of curiosity, how will the time range from the request (which has fields `to` and `from`) be mapped to a string?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-analytics.yaml [hunk 5]
```diff
@@ -1230,11 +1371,274 @@ components:
...
+            required:
+                - type
+                - timeRange
+                - metrics
+            description: |
+                Defines the analytics request associated with a widget. This determines what data the widget displays.
+
+                The `type` field selects the analytics request type:
+                - **measures**: Returns aggregated values. Uses `MetricRequest` for metrics.
+                - **facets**: Groups data by facet values. Uses `FacetMetricRequest` for metrics, and supports `by` and `limit` fields.
```
Comment (carlos-andres-osorio) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-analytics.yaml]: will support for time-series be added in a future PR?

File: gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcCustomDashboardRepository.java [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+            }
+            first = false;
+            builder.append(escapeReservedWord(getDBName(column.name)));
+            builder.append(" = ?");
+        }
+        builder.append(", labels = ?, widgets = ? where id = ?");
+        return builder.toString();
+    }
+
+    private <T> T deserialize(String json, TypeReference<T> typeRef) {
```
Comment (carlos-andres-osorio) [inline gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcCustomDashboardRepository.java:173]: If data can't be deserialized, the dashboard will be returned with empty widgets/labels. Would it be better (more explicit to the user) to return an error?

File: gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcCustomDashboardRepository.java [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+        }
+        try {
+            return JSON_MAPPER.readValue(json, typeRef);
+        } catch (Exception e) {
+            log.error("Failed to deserialize JSON", e);
+            return null;
+        }
+    }
+
+    private String serialize(Object value) {
```
Comment (carlos-andres-osorio) [inline gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcCustomDashboardRepository.java:184]: If data can't be serialized, the dashboard will be saved with empty widgets/labels. Would it be better to return an error?

File: gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcCustomDashboardRepository.java [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.stereotype.Repository;
+
+/**
+ * @author Antoine CORDIER (antoine.cordier at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+@CustomLog
+@Repository
+public class JdbcCustomDashboardRepository
```
Comment (carlos-andres-osorio) [inline gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcCustomDashboardRepository.java:49]: Based on the current scope, do you know if method `findAll()` will be used at any point?

If yes, shouldn't it be overridden here, or dashboards won't be loaded correctly.

File: gravitee-apim-repository/gravitee-apim-repository-test/src/test/java/io/gravitee/repository/management/CustomDashboardRepositoryTest.java [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+import static io.gravitee.repository.utils.DateUtils.compareDate;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.fail;
+
+import io.gravitee.repository.management.model.CustomDashboard;
+import io.gravitee.repository.management.model.CustomDashboardWidget;
+import java.util.*;
+import org.junit.Test;
+
+public class CustomDashboardRepositoryTest extends AbstractManagementRepositoryTest {
```
Comment (carlos-andres-osorio) [inline gravitee-apim-repository/gravitee-apim-repository-test/src/test/java/io/gravitee/repository/management/CustomDashboardRepositoryTest.java:27]: Not an issue in these tests, but tests in this package don't follow a consistent naming convention. 

Is there a preference for using camelCase vs. snake_case with   `DisplayNameGeneration`?

File: gravitee-apim-repository/gravitee-apim-repository-test/src/test/java/io/gravitee/repository/management/CustomDashboardRepositoryTest.java [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+            .build();
+
+        var dashboard = CustomDashboard.builder()
+            .id("new-cd")
+            .organizationId("DEFAULT")
+            .name("New Dashboard")
+            .createdBy("user-new")
+            .createdAt(new Date(1000000000000L))
+            .lastModified(new Date(1111111111111L))
+            .labels(Map.of("env", "prod"))
```
Comment (carlos-andres-osorio) [inline gravitee-apim-repository/gravitee-apim-repository-test/src/test/java/io/gravitee/repository/management/CustomDashboardRepositoryTest.java:115]: are empty labels/widgets treated in a way that should be tested explicitly?

File: gravitee-apim-repository/gravitee-apim-repository-api/src/main/java/io/gravitee/repository/management/model/CustomDashboardWidget.java (hunk 4 above)
Comment (a-cordier) [reply gravitee-apim-repository/gravitee-apim-repository-api/src/main/java/io/gravitee/repository/management/model/CustomDashboardWidget.java]: That's a good point, this was taken from the ts file maybe something is wrong here thought @samirtechlab ?

File: gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcCustomDashboardRepository.java (hunk 8 above)
Comment (a-cordier) [reply gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcCustomDashboardRepository.java:49]: I think the only method we will use will be findByOrganizationId

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-analytics.yaml (hunk 5 above)
Comment (a-cordier) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-analytics.yaml]: good point, it's already supported, I'll remove this entirely, it should just be an internal link or something

File: gravitee-apim-repository/gravitee-apim-repository-test/src/test/java/io/gravitee/repository/management/CustomDashboardRepositoryTest.java (hunk 9 above)
Comment (a-cordier) [reply gravitee-apim-repository/gravitee-apim-repository-test/src/test/java/io/gravitee/repository/management/CustomDashboardRepositoryTest.java:27]: all the tests in the repository being junit 4 tests, I followed the convention here, we use DisplayNameGenerator with junit 5

File: gravitee-apim-repository/gravitee-apim-repository-test/src/test/java/io/gravitee/repository/management/CustomDashboardRepositoryTest.java (hunk 10 above)
Comment (a-cordier) [reply gravitee-apim-repository/gravitee-apim-repository-test/src/test/java/io/gravitee/repository/management/CustomDashboardRepositoryTest.java:115]: Adding a test case for labels, empty/null widgets will be rejected by the service layer

File: gravitee-apim-repository/gravitee-apim-repository-api/src/main/java/io/gravitee/repository/management/model/CustomDashboardWidget.java (hunk 4 above)
Comment (SamHardi) [reply gravitee-apim-repository/gravitee-apim-repository-api/src/main/java/io/gravitee/repository/management/model/CustomDashboardWidget.java]: Timerange is an object as you can see [here](https://github.com/gravitee-io/gravitee-api-management/blob/97d4d90f5632df3cc6400cf75d29d9f2b1298677/gravitee-apim-portal-webui-next/projects/gravitee-dashboard/src/lib/components/widget/model/request/request.ts).

`export interface GlobalRequest<E extends RequestType> {
  type: E;
  timeRange?: TimeRange;
  filters?: RequestFilter[];
  metrics: MetricRequest[];
}`

`export interface TimeRange {
  from: string;
  to: string;
}`

File: gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcCustomDashboardRepository.java (hunk 7 above)
Comment (a-cordier) [reply gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcCustomDashboardRepository.java:184]: It was the pattern in place everywhere, but fair enough, let's throw an illegal state exception :)

File: gravitee-apim-repository/gravitee-apim-repository-api/src/main/java/io/gravitee/repository/management/model/CustomDashboardWidget.java (hunk 4 above)
Comment (a-cordier) [reply gravitee-apim-repository/gravitee-apim-repository-api/src/main/java/io/gravitee/repository/management/model/CustomDashboardWidget.java]: done

**Review by carlos-andres-osorio — APPROVED** [review]
